### PR TITLE
Add Boussinesq approximation to INS

### PR DESCRIPTION
**Pull request by lindsayad**
The result below shows an incompressible Navier Stokes simulation in which the right wall is held at 400K while the left wall is held at 300K. The gravity vector points down.

boussinesq

How about that, hot air rises, and cold air sinks!

This kernel was originally written by @gridley in Moltres. It has been adapted to fit into the new AD, vector variable INS implementation.

@YaqiWang. @fdkong may also find this interesting

### CODE REVIEW

**Comment by lindsayad** [comment]
@jiangwen84 I dunno whether this might be useful to your additive manufacturing work?

**Comment by jiangwen84** [comment]
> @jiangwen84 I dunno whether this might be useful to your additive manufacturing work?

Yes, actually I have included this term in my extra moment source term. I will remove it from my code and use yours instead. Thanks.

**Comment by YaqiWang** [comment]
Yes, this is interesting to me. Once this is in, I will try this to set up one of the MSR benchmarks.

File: modules/navier_stokes/include/userobjects/INSADObjectTracker.h [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+ */
+class INSADObjectTracker : public GeneralUserObject
+{
+public:
+  static InputParameters validParams()
+  {
+    InputParameters params = GeneralUserObject::validParams();
+    params.addClassDescription("User object used to track the kernels added to an INS simulation "
+                               "and determine what properties to calculate in INSADMaterial");
+    return params;
```
Comment (jiangwen84) [inline modules/navier_stokes/include/userobjects/INSADObjectTracker.h:36]: You might want to set 'execute_on' = INITIAL?

Comment (lindsayad) [reply modules/navier_stokes/include/userobjects/INSADObjectTracker.h:36]: That's fair, although the `execute` method is empty

**Comment by jiangwen84** [comment]
@lindsayad my two cents:  I think  ``INSADObjectTracker`` is good because the it ensures to include all the terms into the strong residual. However, it becomes cumbersome for developers to add additional term to original NS equation as they have to change this user object class.

**Comment by lindsayad** [comment]
I thought about that, but I guess I'm always thinking in the context of stabilization. What is more cumbersome, having to add a corresponding `INSAD*XYZ*SUPG/PSPG*Kernel` every time you add a new kernel, or having to add a few things to the object tracker? I'm open to design suggestions

**Comment by YaqiWang** [comment]
I think you should have an action for adding NS kernels, materials, etc.

The input syntax can be simply as
```
[NavierStokes]
  block = '0 1'
  velocity_variable = 'velocity'
  pinned_node = pinned_node
  boussinesq_approximation = true
  dirichlet_boundary_condition = 'left bottom right'
  density =  'rho'
  viscosity =  'mu'
  heat_capacity = 'cp'
  thermal_conductivity =  'k'
  # something like this
[]
```
Users does not need to add those object directly. I did not look into `INSADObjectTracker` but guess you do not need it if you have the action.

**Comment by lindsayad** [comment]
The user doesn't have to add the `INSADObjectTracker` either. It happens
for them automatically. Wen's two cents are related to developers

**Comment by YaqiWang** [comment]
You misunderstood me. I mean you do not need that uo at all.

**Comment by jiangwen84** [comment]
@lindsayad I write a new INSMaterial that derives from INSADMaterial and add extra residual term to existing one, so I reuse your PSPS/SUPG kernels.

**Comment by lindsayad** [comment]
> @lindsayad I write a new INSMaterial that derives from INSADMaterial and add extra residual term to existing one, so I reuse your PSPS/SUPG kernels.

So it seems to me that the design I've created here should work for you right? Do you call `INSADMaterial::computeQpProperties` in your derived class? Then you could easily add to the `_momentum_strong_residual` after that

**Comment by lindsayad** [comment]
@YaqiWang with respect to the action, I (or somebody else if they beat me to it) will definitely write an incompressible Navier Stokes action at some point. Even when the action gets created, however, I would like for users to be able to setup clean input files without using the action. I really don't like it when I come across modules/applications in which you can't even setup the physics without using an action. I feel like I have no granular control at that point. MOOSE was designed to be modular.

I really like the user object (which gets created automatically without the user having to specify anything in the input file) because it eliminates duplication of parameters. For instance because of stabilization, both `INSADMomentumPressure` and `INSADMaterial` need to know the value of the parameter `integrate_p_by_parts`. Without the user object, the user has to make sure that they consistently set that parameter in the input file, whether specifying it twice in the different blocks, or by specifying it once in the `GlobalParams` block (and I know how much you like `GlobalParams`). Having to specify multiple times is prone to error, and even I myself have slipped up before. With the user object, the parameter is specified one time naturally in the `INSADMomentumPressure` kernel block, it's fed into the user object, and then `INSADMaterial` picks up its value through the user object. 

I also used to have to specify a `transient_term` boolean to `INSADMaterial` that said whether a transient component needed to be included in the stabilization parameter `tau`. However, now once a user specifies an `INSADMomentumTimeDerivative` kernel in the input file, the transient term calculation is automatically triggered in the material. This would have been another instance where the user could have made a mistake in the input.

I recognize that the issues raised in the above two paragraphs would also be fixed with an action, but I reassert that I want to continue to enable input file modularity and fine-grained user control. (Also sometimes a user may simply want to add pieces by hand so that they understand what's going on; I know that this is not the more common case)

**Comment by gridley** [comment]
I'd like to add that the reference temperature is not be necessary, since we're only looking at differential operators acting on T in this case. Numerically, they're not strictly equivalent if the temperature Laplacian is large, but a Boussinesq buoyancy approximation won't be valid in that case anyways for any fluid with a reasonable thermal expansivity. Might be worth considering removing that.

**Comment by jiangwen84** [comment]
> > @lindsayad I write a new INSMaterial that derives from INSADMaterial and add extra residual term to existing one, so I reuse your PSPS/SUPG kernels.
> 
> So it seems to me that the design I've created here should work for you right? Do you call `INSADMaterial::computeQpProperties` in your derived class? Then you could easily add to the `_momentum_strong_residual` after that

Yes, you are correct. For the time being, I could not think of a better design than your UO.

**Comment by lindsayad** [comment]
@gridley I'll let you remove it when you start your internship :-)

**Comment by YaqiWang** [comment]
@lindsayad Appreciate your work here. It is great. But I think having an action will make it greater. Note that MOOSE can automatically convert actions into a full input if I remember correctly (added by Daniel? @dschwen). As a developer, I like modularity and fine-grained user control as well. But the action will play a big role for a user, an expert for example in reactor analysis, who does not know much fluid but is required to coupling fluid. It simplify their life a lot. Your UO removes the issue related with parameter duplication, which is great, but still requires users to have some more knowledge on setting up kernels, where the users might miss one of the kernels. I hate people to be not considerate. Just my cents.

**Comment by YaqiWang** [comment]
I might spend some time on adding the action if you want. It should be only few hundred lines, not a big deal. I of course need to educate myself a little on the physics beforehand.

**Comment by lindsayad** [comment]
> added by Daniel? @dschwen).

Yea `DumpObjectsProblem`

> But the action will play a big role for a user, an expert for example in reactor analysis, who does not know much fluid but is required to coupling fluid. It simplify their life a lot. Your UO removes the issue related with parameter duplication, which is great, but still requires users to have some more knowledge on setting up kernels, where the users might miss one of the kernels. I hate people to be not considerate. Just my cents.

Yes I was not advocating against an action; I was just advocating for the UO which made it so that it's difficult for a user to make an error if they *don't* use an action.

> I might spend some time on adding the action if you want. It should be only few hundred lines, not a big deal. I of course need to educate myself a little on the physics beforehand.

I've created #15159 to track this. It would be awesome if you added it. One of us will get to it at some point.

**Comment by permcody** [comment]
I've read through the comments and I think this PR is good enough to stand as-is. Adding the Action is a good idea, but shouldn't hold this up. I tend to agree with Alex that it's handy to modularize things when we can, but I understand different developers have different views.

**Review by permcody — APPROVED** [review]

**Comment by lindsayad** [comment]
> I'd like to add that the reference temperature is not necessary, since we're only looking at differential operators acting on T in this case. Numerically, they're not strictly equivalent if the temperature Laplacian is large, but a Boussinesq buoyancy approximation won't be valid in that case anyways for any fluid with a reasonable thermal expansivity. Might be worth considering removing that.

I've been thinking about this because of [this discussion on moose-users](https://groups.google.com/d/msgid/moose-users/2de9c60e-e06f-4341-976e-c13d1d9d90deo%40googlegroups.com?utm_medium=email&utm_source=footer). How does the reference temperature *not* matter? If I had a uniform temperature of say 300 down a channel, if I alternately picked a reference temperature of 400 and then 200, wouldn't I be accelerating the fluid in different directions?!

**Comment by lindsayad** [comment]
I’m feeling like a crazy person...a uniform temperature field shouldn’t drive flow at all...so what gives?!

**Comment by jiangwen84** [comment]
Alex, I have not read the details in that discussion.  How does that issue relate to Boussinesq implementation here?

**Comment by gridley** [comment]
I see what you’re saying Alex. I think it depends on the BCs. If it’s all free-stream, then you should see a difference. If it’s a closed boundary, it should end up giving larger pressures for larger delta T’s, but the pressure gradients should have the same values.

I can’t come up with a good proof for this at the moment.

**Comment by lindsayad** [comment]
The user has what looks like fairly reasonable dimensions and material coefficients, but the forces related to gravity are very large. I had to scale the gravity vector down by 9 orders of magnitude in order to get his solve to converge. So I’m trying to understand where the seemingly huge magnitude of these forces are coming from.

**Comment by gridley** [comment]
Hm. I tried working through this on paper and realized I’m probably wrong! I’ll have to play with the NS module a bit soon to figure this out. I’ve been meaning to anyways.

**Comment by lindsayad** [comment]
Ok to close the loop on this, the choice of reference temperature *does not* have any impact on the velocity solution (and consequently no impact on the temperature solution), exactly as @gridley postulated. This is independent of steady or transient simulation. This is because the pressure (really just a Lagrange Multiplier) adjusts itself such that continuity of mass condition is satisfied. So choice of reference temperature does have a *huge* impact on the pressure field. Using @gridley 's `boussinesq_square` example:

## Reference temperature = 900
### pressure
p-900
### temperature
temp-900
### velocity magnitude
vel-900

## Reference temperature = 0
### pressure
p-0
### temperature
temp-0
### velocity magnitude
vel-0

I was thinking about this since @fdkong is working on a natural convection problem.

**Comment by gridley** [comment]
For the case of all no-slip BCs, I think you're right. But I get the feeling that if there are any inlet or outlet conditions with fixed pressure, you will of course see big differences in the solution.

**Comment by lindsayad** [comment]
You are right indeed. You don't even need explicit boundary conditions on the pressure to generate differences in the solution. E.g. if you integrate the pressure term by parts and apply an integrated boundary condition for the boundary-normal velocity anywhere, then you will see solution differences. So essentially if p appears anywhere in a boundary condition other than a single-dof pressure pin to make your system nonsingular (which isn't really a boundary condition anyways), then you will see different velocity solutions for different reference temperatures

**Comment by lindsayad** [comment]
So I think the moral of the story is that for real physics simulations, you should most certainly use an informed reference temperature, e.g. a T0 corresponding to rho0

**Pull request merged**